### ISSUE TITLE
Intermittent onComplete observed without onNext, response content dropped

### ISSUE THREAD

**Issue by danielra** [jetty-project/jetty-reactive-httpclient#14]
I am using the jetty-reactive-httpclient via the Spring 5 WebClient/JettyClientHttpConnector in a Spring WebFlux service, and I see that intermittently the body content from a response is not emitted as expected - instead I see an onComplete call without a prior onNext. However, the debug log lines from the jetty-client indicate that the response body _was_ received/decoded. I seem to be able to reproduce the issue more easily if my machine is busy using lots of CPU / cores in the background - so I suspect some sort of race condition or other multi-threading issue is behind this unexpected behavior. Even with CPU busy in the background, it still usually takes many tries to reproduce - perhaps something like 1 in every 100 tries. I am not sure that the problem is in jetty-reactive-httpclient code, but some of the relevant/potentially interesting debug log lines come from here so I am hoping at least for some ideas/guidance.

I have captured debug logging for a request that encountered this problem, as well as debug logging for the same sort of request that worked as expected for comparison. One detail that seems potentially significant is the name of the thread which various log lines origniate from. In the normal "working as expected" case only the first few log lines come from the WebFlux/tomcat event loop thread named like "http-nio-16027-exec-#", whereas when the problem occurs some of the ResponseListenerPublisher messages etc. also end up coming from this thread rather than from a thread named like "HttpClient*" as in the normal "working as expected" case.
I've attached the captured debug log output:
[jetty_reactive_webclient_problem_log.txt](https://github.com/jetty-project/jetty-reactive-httpclient/files/2842557/jetty_reactive_webclient_problem_log.txt)
[jetty_reactive_webclient_normal_log.txt](https://github.com/jetty-project/jetty-reactive-httpclient/files/2842558/jetty_reactive_webclient_normal_log.txt)

It can be seen in the normal log that after the response content is decoded there is an onNext call made prior to onComplete. However, in the problem log after the response content is decoded, there is an onComplete call but no prior onNext call - and it is then subsequently logged by the default reactor operators that there was an onNextDropped with a DefaultDataBuffer (possibly leaked?).

For reference, the client I am using is essentially instantiated like:
```
final HttpClient httpClient = new HttpClient(new SslContextFactory(true));
httpClient.setAddressResolutionTimeout(httpConfig.getConnectionRequestTimeout());
httpClient.setConnectTimeout(httpConfig.getConnectionRequestTimeout());
httpClient.setIdleTimeout(httpConfig.getSocketTimeout());
httpClient.setTCPNoDelay(true);
httpClient.setMaxConnectionsPerDestination(httpConfig.getMaxTotalConnections());
httpClient.setFollowRedirects(false);
httpClient.start();

final ClientHttpConnector clientHttpConnector = new JettyClientHttpConnector(httpClient);
final WebClient webClient = WebClient.builder()
        .clientConnector(clientHttpConnector)
        .build();
```

The WebClient is then being used like:
```
return client.get()
        .uri(uri)
        .retrieve()
        .bodyToMono(ResponseModel.class)
        .log() // You will see the events such as onSubscribe / onComplete in the attatched/linked log files emitted due to this Mono.log call
        .map(this::processResult)
...
```

Relevant library versions my project is using are:
`org.eclipse.jetty:jetty-reactive-httpclient:1.0.2`
`org.eclipse.jetty:jetty-client:9.4.14.v20181114`
And using version `5.1.4.RELEASE` of the spring framework.

Given the intermittent nature of the problem, I haven't been able to produce a simple test case that reliably reproduces the issue. Please do let me know if any further information from my end would be useful. Any ideas/guidance will be appreciated. In the interim, I will keep digging into the relevant library source code to see if I can better understand the problem and/or propose a fix.

Thank you!

**Comment by kptfh** [comment]
Just curious,  why did you select JettyClientHttpConnector instead of default Netty based?
Does it fail on default Netty based http connector as well?
I suppose that there is some issue in JettyClientHttpConnector.

Try to use [https://github.com/Playtika/feign-reactive](https://github.com/Playtika/feign-reactive) 
It has reactive jetty HttpClient based implementation
[https://github.com/Playtika/feign-reactive/tree/develop/feign-reactor-jetty](https://github.com/Playtika/feign-reactive/tree/develop/feign-reactor-jetty)

**Comment by danielra** [comment]
I switched to using the JettyClientHttpConnector because I haven't been able to get the reactor-netty based client to timeout requests without leaking Netty's pooled ByteBufs. They have been working to resolve these sorts of leaks for a while and have fixed a number of them - but I still see a lot of leak reports with the newest version of the client (I plan to file an issue over there once I have time to set up another isolated repro case).

Thanks for the suggestion, I'll take a look at that as another alternative option.

**Comment by sbordet** [comment]
@danielra here's my analysis from your "problem" log:

```
2019-02-06T12:58:06.528-08:00 [HttpClient@1821418353-70][o.e.j.r.c.i.QueuedSinglePublisher] DEBUG: offered item ContentChunk@40f05073 to QueuedSinglePublisher@412aed24
```
We see the response content and we produce a `ContentChunk`, which gets queued.
There is no demand yet, so this particular chunk is not processed immediately by us (in the "normal" case there already is demand at this point).

```
2019-02-06T12:58:06.531-08:00 [http-nio-16027-exec-9][o.e.j.r.c.i.QueuedSinglePublisher] DEBUG: QueuedSinglePublisher@412aed24 subscription from reactor.core.publisher.FluxMap$MapSubscriber@644f2a52
2019-02-06T12:58:06.531-08:00 [http-nio-16027-exec-9][o.e.j.r.c.i.QueuedSinglePublisher] DEBUG: processing next item ContentChunk@40f05073 by QueuedSinglePublisher@412aed24
2019-02-06T12:58:06.531-08:00 [http-nio-16027-exec-9][o.e.j.c.HttpReceiver] DEBUG: Content consumed asynchronously, resuming processing
```
The first line creates the demand; in the second line we process the `ContentChunk` - which means that we call `onNext()`; the third line demonstrates that the "application" we called in `onNext()` succeeded the `ContentChunk` callback.
Here you see there are 2 layers of "application": we call Reactor, which in turn calls your application code.
We called Reactor properly, but Reactor did not call your application yet - I don't know why, perhaps there is no demand?

Because Reactor succeeded the `ContentChunk` callback, we resume the processing of the response, which is complete and produces:
```
2019-02-06T12:58:06.533-08:00 [http-nio-16027-exec-9][o.e.j.r.c.i.QueuedSinglePublisher] DEBUG: completed QueuedSinglePublisher@412aed24
2019-02-06T12:58:06.533-08:00 [http-nio-16027-exec-9][o.e.j.r.c.i.QueuedSinglePublisher] DEBUG: processing last item org.eclipse.jetty.reactive.client.internal.QueuedSinglePublisher$$Lambda$437/1821616361@114dfbf by QueuedSinglePublisher@412aed24
2019-02-06T12:58:06.533-08:00 [http-nio-16027-exec-9][h.g.c.6.3.6] INFO: | onComplete()
```
The second line invokes `onComplete()` on Reactor, which immediately calls your application.

From our point of view, we call Reactor properly; seems that Reactor is not doing the right thing towards your application.

We have been collaborating with the Reactor guys so if you file an issue there, link this issue and they'll have a better context to work on.

**Comment by sbordet** [comment]
Also, I'm no Reactor expert: could it be that your use of the Reactor APIs is wrong?

**Comment by danielra** [comment]
Thank you very much for taking the time to provide your analysis. I am working towards a more isolated/complete example which reproduces the problem. I will follow up on the reactor side and link this issue as you suggest.

I am not a Reactor expert either, and I am certainly open to learning that I am doing something wrong on my end. In this case, the actual subscription / demand is coming from Spring WebFlux framework code - it seems also possible there could be an issue there. I'll keep digging. Thanks again!

**Comment by danielra** [comment]
While paring down my actual project into something more minimal/isolated to reproduce the issue, I started seeing a new NPE error occurring in reactor code. I reported this issue over on the reactor project here: https://github.com/reactor/reactor-core/issues/1522 . Someone there noted that it appeared as though onNext is being called by Jetty code after onComplete has already been called. A potential issue with continuing a loop after observing a terminal event was noted - see their original comment here: https://github.com/reactor/reactor-core/issues/1522#issuecomment-462670538

I also produced a demo project with which I am able to reproduce the issue. Details on that, and new example debug logs can be found here: https://github.com/reactor/reactor-core/issues/1522#issuecomment-462887043

**Comment by sbordet** [comment]
@danielra can you try the latest code? I pushed a fix for this behavior.

**Comment by danielra** [comment]
@sbordet I built the current master branch and used that version locally and I am happy to report that I am no longer able to reproduce the issue I was seeing. Thank you very much for your work here - I appreciate it! Looking forward to an official release which includes the update. :)

**Comment by sbordet** [comment]
@danielra a new release of this project is scheduled just after Jetty 9.4.15, which should be due early next week. Thanks for the feedback on the fix!

**Comment by sbordet** [comment]
I have released version 1.0.3.
